tests/api/test_utils.py

```diff
@@ -108,6 +108,36 @@ def some_function(page_size=8, page_offset=0):
         self.assertEqual((4 * page_size - 1) * ["a"], some_list)
         self.assertEqual(len(some_list), (4 * page_size - 1))
 
+    def test_paginate_endpoint__string(self):
+        def some_function(page_size=8, page_offset=0):
+            if page_offset > 3 * page_size:
+                assert False  # should not happen
+            elif page_offset > 2 * page_size:
+                return (page_size - 1) * "a"
+            else:
+                return page_size * "a"
```

Comment:
In my eyes, this is quite hard to understand. Perhaps add a comment `Assuming that there are 2*page_size - 1 items.` Or directly create it differently and allow to merge all the different tests into one parametrised test:

```
class PaginatableSequence:
    def __init__(self, n_items: int):
        self.items = [i for i in range(n_items)]
        
    def call(self, page_offset: int, page_size: int):
        if page_offset >= len(self.items):
            return []
        if page_offset + page_size > len(self.items):
            return self.items[page_offset:]
        return self.items[page_offset:page_offset+page_size]
```

With such a design, you could also test all the cases easily:
- n_items < page_size
- n_items = page_size
- n_items > page_size and n_items < 2* page_size
- n_items = 2 *page_size
- n_items > 2 * page_size.

Furthermore, you could 
```
some_iterator = paginate_endpoint(paginatable_sequence.call)
some_list == list(some_iterator)
assert some_list == paginatable_sequence.items
```

With a spy, you could also do
```
paginatable_sequence_call_spy.assert_has_calls([
  call(page_offset=0, page_size=8),
  ...
])
```
Asserting the calls directly is also better than having some `assert False # should not happen` in some cases.

Reply:
I don't see how this has been addressed apart from the comments being better.

Reply:
> In my eyes, this is quite hard to understand. Perhaps add a comment Assuming that there are 2*page_size - 1 items

I made this change. I think the tests are more verbose the way they are than what you proposed.

Reply:
I don't think that they are super good, but ok enough.